ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/attestation/MatchingDataAttestationGroupV2.java

```diff
@@ -0,0 +1,520 @@
+                    singleAttestationsByCommitteeIndex.get(committeeIndex);
+                if (singleAttestations != null) {
+                  singleAttestations.removeIf(sa -> attestation.bits().isSuperSetOf(sa.bits()));
+                }
+              });
+    }
+
+    return added;
+  }
+
+  /**
+   * Iterates through the aggregation of attestations in this group. The iterator attempts to create
+   * the minimum number of attestations that include all attestations in the group.
+   *
+   * <p>committeeIndex is an optional parameter that enables aggregation over a specified committee
+   * (applies to Electra only)
+   *
+   * <p>While it is guaranteed that every validator from an attestation in this group is included in
+   * an aggregate produced by this iterator, there is no guarantee that the added attestation
+   * instances themselves will be included.
+   *
+   * @return an iterator including attestations for every validator included in this group.
+   */
+  private Iterator<PooledAttestation> createAggregatingIterator(
+      final long timeLimitNanos,
+      final Supplier<Stream<PooledAttestation>> candidatesStreamSupplier) {
+    readLock.lock();
+    try {
+      // Capture a copy of includedValidators under lock for the iterator's isolated use
+      AttestationBits includedValidatorsCopy = this.includedValidators.copy();
+
+      final AggregatingIterator iterator =
+          new AggregatingIterator(includedValidatorsCopy, candidatesStreamSupplier);
+      if (timeLimitNanos == Long.MAX_VALUE) {
+        return iterator;
+      }
+
+      return new TimeLimitingIterator<>(
+          nanosSupplier,
+          timeLimitNanos,
+          iterator,
+          __ -> LOG.info("Time limit reached, skipping aggregation"));
+    } finally {
+      readLock.unlock();
+    }
+  }
+
+  public Stream<PooledAttestationWithData> streamForBlockProduction(final long timeLimitNanos) {
+    return StreamSupport.stream(
+            spliterator(timeLimitNanos, blockProductionCandidatesStreamSupplier()), false)
+        .map(
+            pooledAttestation -> new PooledAttestationWithData(attestationData, pooledAttestation));
+  }
+
+  public Stream<PooledAttestationWithData> streamForAggregationProduction(
+      final Optional<UInt64> committeeIndex, final long timeLimitNanos) {
+    checkArgument(
+        committeeIndex.isPresent() || !includedValidators.requiresCommitteeBits(),
+        "Committee index must be present if committee bits are required");
+    return StreamSupport.stream(
+            spliterator(
+                timeLimitNanos, aggregationProductionCandidatesStreamSupplier(committeeIndex)),
+            false)
+        .map(
+            pooledAttestation -> new PooledAttestationWithData(attestationData, pooledAttestation));
+  }
+
+  public Stream<PooledAttestationWithData> streamForApiRequest(
+      final Optional<UInt64> committeeIndex, final boolean requiresCommitteeBits) {
+    boolean noMatch;
+    readLock.lock();
```

Comment:
we should be keeping the lock to where we use the data - so in this case this lock should be inside `noMatchingAttestations`
this improves readabilty but also centralises the lock so we know any time we call that fn its correct, otherwise if we ever use it elsewhere we'd also need locks there too